File: ios/BiblosSurfer/DataModels/StoredModels.swift
//
//  StoredModels.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 21/08/2026.
//

import Foundation
import SwiftData

@Model
final class StoredBook {
    var id: UUID
    var fileName: String
    var title: String
    var author: String?
    var relativePath: String
    var coverFileName: String?
    var locatorJSON: String?
    var progression: Double
    var isPDF: Bool
    var folderName: String?
    var importedAt: Date
    @Relationship(deleteRule: .cascade, inverse: \StoredBookmark.book)
    var bookmarks: [StoredBookmark]

    init(
        id: UUID = UUID(),
        fileName: String,
        title: String,
        author: String? = nil,
        relativePath: String,
        coverFileName: String? = nil,
        locatorJSON: String? = nil,
        progression: Double = 0,
        isPDF: Bool,
        folderName: String? = nil,
        importedAt: Date = Date(),
        bookmarks: [StoredBookmark] = []
    ) {
        self.id = id
        self.fileName = fileName
        self.title = title
        self.author = author
        self.relativePath = relativePath
        self.coverFileName = coverFileName
        self.locatorJSON = locatorJSON
        self.progression = progression
        self.isPDF = isPDF
        self.folderName = folderName
        self.importedAt = importedAt
        self.bookmarks = bookmarks
    }
}

@Model
final class StoredFolder {
    var name: String

    init(name: String) {
        self.name = name
    }
}

@Model
final class StoredBookmark {
    var locatorJSON: String
    var createdAt: Date
    var title: String?
    var book: StoredBook?

    init(locatorJSON: String, createdAt: Date = Date(), title: String? = nil, book: StoredBook? = nil) {
        self.locatorJSON = locatorJSON
        self.createdAt = createdAt
        self.title = title
        self.book = book
    }
}

File: android/app/src/main/java/miko/biblossurfer/data/StoredModels.kt
package miko.biblossurfer.data

import androidx.room.Dao
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Relation
import androidx.room.Transaction
import androidx.room.TypeConverter
import androidx.room.TypeConverters
import androidx.room.Update
import miko.biblossurfer.data.model.PublicationFormat
import java.util.Date
import java.util.UUID

@Entity(tableName = "stored_books", indices = [Index(value = ["relativePath"], unique = true)])
data class StoredBook(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val fileName: String,
    val title: String,
    val author: String? = null,
    val relativePath: String,
    val coverFileName: String? = null,
    val locatorJSON: String? = null,
    val progression: Double = 0.0,
    val format: PublicationFormat = PublicationFormat.UNKNOWN,
    val folderName: String? = null,
    val importedAt: Long = Date().time,
)

@Entity(tableName = "stored_folders")
data class StoredFolder(
    @PrimaryKey val name: String,
)

@Entity(
    tableName = "stored_bookmarks",
    foreignKeys = [
        ForeignKey(
            entity = StoredBook::class,
            parentColumns = ["id"],
            childColumns = ["bookId"],
            onDelete = ForeignKey.CASCADE,
        )
    ],
    indices = [Index("bookId")]
)
data class StoredBookmark(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val locatorJSON: String,
    val createdAt: Long = Date().time,
    val title: String? = null,
    val bookId: String? = null,
)

class PublicationFormatConverters {
    @TypeConverter
    fun fromFormat(value: PublicationFormat): String = value.name.lowercase()

    @TypeConverter
    fun toFormat(value: String): PublicationFormat =
        PublicationFormat.entries.firstOrNull { it.name.equals(value, ignoreCase = true) }
            ?: PublicationFormat.UNKNOWN
}

@Dao
interface BookDao {
    @Query("SELECT * FROM stored_books WHERE relativePath = :relativePath LIMIT 1")
    fun bookByRelativePath(relativePath: String): StoredBook?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun upsert(book: StoredBook)

    @Update
    fun update(book: StoredBook)

    @Insert
    fun insertBookmark(bookmark: StoredBookmark)
}
